crates/runtime/src/datafusion/extension/bytes_processed.rs

```diff
@@ -285,27 +333,106 @@ impl ExecutionPlan for BytesProcessedExec {
             )
         };
 
-        let bytes_processed_stream = stream! {
-            let mut bytes_processed = 0u64;
-            while let Some(batch) = stream.next().await {
-                match batch {
-                    Ok(batch) => {
-                        bytes_processed += batch.get_array_memory_size() as u64;
-                        yield Ok(batch)
-                    }
-                    Err(e) => {
-                        yield Err(e);
-                    }
-                }
-            }
-            crate::metrics::telemetry::track_bytes_processed(
-                bytes_processed,
-                &request_context.to_dimensions(),
-            );
-        };
+        let bytes_processed_stream = BytesProcessedStream::new(stream, request_context);
```

Comment:
It might be a bit more terse to just use `fold`:

```rust
use futures::stream::{StreamExt, Stream};

let bytes_processed_stream = {
    let request_context_clone = Arc::clone(&request_context);
    stream.fold(0u64, |bytes_processed, result| async move {
        match result {
            Ok(batch) => bytes_processed + batch.get_array_memory_size() as u64,
            Err(_) => bytes_processed,
        }
    })
    .into_stream()
    .map(move |bytes_processed| {
        crate::metrics::telemetry::track_bytes_processed(
            bytes_processed,
            &request_context_clone.to_dimensions(),
        );
        None
    })
    .chain(stream.map(|result| match result {
        Ok(batch) => Ok(batch),
        Err(e) => Err(e),
    }))
};
```